multicluster/controllers/multicluster/node_controller.go

```diff
@@ -73,67 +78,215 @@ func NewNodeReconciler(
-		Name:      node.Name,
-		Namespace: r.namespace,
-	}
+	_, hasGWAnnotation := node.Annotations[common.GatewayAnnotation]
+	nodeIsReady := isReadyNode(node)
+	noActiveGateway := r.activeGateway == ""
+	isActiveGateway := r.activeGateway == node.Name
 
-	// TODO: cache might be stale. Need to revisit here and other reconcilers to
-	// check if we can improve this with 'Owns' or other methods.
-	var gwIP, internalIP string
-	if isGW {
-		if internalIP, gwIP, err = r.getGatawayNodeIP(node); err != nil {
-			klog.ErrorS(err, "There is no valid Gateway IP for Node, will retry later when there is any new Node update", "node", node.Name)
+	if !nodeIsReady {
+		if isActiveGateway {
+			if err := r.recreateActiveGateway(ctx, gw); err != nil {
+				return ctrl.Result{}, err
+			}
 			return ctrl.Result{}, nil
 		}
+		if hasGWAnnotation {
+			r.gatewayCandidates[req.Name] = req.Name
+		}
+		return ctrl.Result{}, nil
 	}
-	if err := r.Client.Get(ctx, gwNamespacedName, gw); err != nil {
-		if apierrors.IsNotFound(err) && isGW {
-			gw.GatewayIP = gwIP
-			gw.InternalIP = internalIP
-			if err := r.Client.Create(ctx, gw, &client.CreateOptions{}); err != nil {
+
+	var err error
+	var gwIP, internalIP string
+	// When Node is ready without Gateway annotation
+	if !hasGWAnnotation {
+		if isActiveGateway {
+			// The annotation is removed from the Node, we need recreate a new Gateway.
+			if err := r.recreateActiveGateway(ctx, gw); err != nil {
 				return ctrl.Result{}, err
 			}
-			return ctrl.Result{}, nil
 		}
-		return ctrl.Result{}, client.IgnoreNotFound(err)
+		return ctrl.Result{}, nil
 	}
 
-	if !isGW {
-		err := r.Client.Delete(ctx, gw, &client.DeleteOptions{})
-		return ctrl.Result{}, client.IgnoreNotFound(err)
+	// When the Node is ready with the Gateway annotation
+	r.gatewayCandidates[node.Name] = node.Name
+	internalIP, gwIP, err = r.getGatawayNodeIP(node)
+	if err != nil {
+		klog.ErrorS(err, "There is no valid Gateway IP for Node", "node", node.Name)
+		if isActiveGateway {
+			if err := r.recreateActiveGateway(ctx, gw); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+		return ctrl.Result{}, nil
 	}
-
-	gw.GatewayIP = gwIP
 	gw.InternalIP = internalIP
-	if err := r.Client.Update(ctx, gw, &client.UpdateOptions{}); err != nil {
-		return ctrl.Result{}, err
+	gw.GatewayIP = gwIP
+	if isActiveGateway {
+		if err := r.updateActiveGateway(ctx, gw); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+	if noActiveGateway {
+		if err := r.createGateway(gw); err != nil {
+			return ctrl.Result{}, err
+		}
 	}
 	return ctrl.Result{}, nil
 }
 
+// initializeActiveGateway initializes 'activeGateway' and 'gatewayCandidates'
+// during controller startup.
+func (r *NodeReconciler) initializeActiveGateway() error {
+	ctx := context.Background()
+	nodeList := &corev1.NodeList{}
+	if err := r.Client.List(ctx, nodeList, &client.ListOptions{}); err != nil {
+		return err
+	}
+	// Empty the cache in case of initializeActiveGateway failed and rerun.
```

Comment:
failure -> failure

Reply:
I do not see you added code to do the cleanup? Maybe simpler to move the code to the end of func before returning without no error.

Reply:
Done, add the candidates initialization before return nil.